### PR TITLE
Api methods for getting data by index

### LINKED ISSUE #414: Add API to get chart values (data, markers, etc)
Why is there no getPrice?
or getMarkers();

Would really like them, makes it much easyer for me to use.

Example:
This is how I need to get the current price now:
        `let currentPrice = Array.from(Array.from(priceArea._dataUpdatesConsumer.de.ee)[priceArea._dataUpdatesConsumer.de.ee.size - 1][1].mapping)[0][1].value;`

Just write getPrice(); would make it much more easyer and would have saved me at least 3 hours of searching.

### LINKED ISSUE #470: Fix types inconsistency on API level with time
Currently we have several "form" of time:

- inbound time - it might be "business day string", "business day object" and "utc timestamp"
- outbound time - even if we declared it as `Time` (what means the same as "inbound time"), but actually it might be "business day object" and "utc timestamp" only, because we strip strings on API level and convert them in "business day object"
- "time comes from model level" - that times, which are generated on model level, which knows nothing about `Time` and knows about "business day object" and "utc timestamp" (the same as outbound one, but more explicit on type level)

I think we need to make it more consistency and leave the only one possible time value (whatever it will be) (or maybe two of them, but split them out and make somehow).

In internal discussion @eugene-korobko suggests to use "inbound time" everywhere, and provide the same time value an user passed to the library (currently we parse/convert it and just forget about the original value). But there is question "what we need to do if we have 2 series, one has data with "business day object" and another one with "business day string", and they have the same time point, what value we should provide here? Which was set first?

Me personally thinks that it might be better to split explicitly "inbound" and "outbound" types and use them (actually it's pretty the same we currently have, but without having wrong types somewhere).

### PR DESCRIPTION
**Pull request by kirchet**
**Type of PR:** enhancement

**PR checklist:**

- [x] Addresses an existing issue: fixes #414, fixes #470 
- [x] Includes tests
- [x] Documentation update

**Overview of change:**

Added methods

- `ISeriesApi.getMarkers`
- `ISeriesApi.dataByIndex`

Changed time types everywhere in the public API to `Time`

### CODE REVIEW

**Review by timocov — CHANGES_REQUESTED** [review]

File: src/api/iseries-api.ts [hunk 1]
```diff
@@ -158,6 +159,19 @@ export interface ISeriesApi<TSeriesType extends SeriesType> {
...
+	 *
+	 * @param index - Logical index
+	 * @param mismatchDirection - Search direction if no data found at provided logical index.
+	 * @returns Original data item provided via setData or update methods.
+	 * @example
+	 * ```js
+	 * const originalData = series.dataByIndex(10, LightweightCharts.MismatchDirection.NearestLeft);
+	 * ```
+	 */
+	dataByIndex(index: Logical, mismatchDirection?: MismatchDirection): SeriesDataItemTypeMap[TSeriesType] | null;
```
[the remark this answers is not in the document]
Comment (kirchet) [reply src/api/iseries-api.ts]: https://github.com/tradingview/lightweight-charts/pull/934/commits/3353c9fc2d22731913a10b0a39eab3c97216dfa2

File: src/model/plot-data.ts [hunk 2]
```diff
@@ -21,4 +21,6 @@ export interface PlotRow {
 	readonly index: TimePointIndex;
 	readonly time: TimePoint;
 	readonly value: PlotRowValue;
+	readonly originalTime: OriginalTime;
 }
+export type WhitespacePlotRow = Omit<PlotRow, 'value'>;
```
[the remark this answers is not in the document]
Comment (kirchet) [reply src/model/plot-data.ts]: https://github.com/tradingview/lightweight-charts/pull/934/commits/f66ff411f50ad31585a491930410ca2f63f32436#diff-219c330dbd21b6f81cb7dc9b0945406461a0100fb85baa27ec048d621031e3d9R26-R27

File: src/api/series-api.ts [hunk 3]
```diff
@@ -126,16 +127,42 @@ export class SeriesApi<TSeriesType extends SeriesType> implements ISeriesApi<TSe
...
 		}));
 		this._series.setMarkers(convertedMarkers);
 	}
 
+	public markers(): readonly SeriesMarker<Time>[] {
+		return this._series.markers().map((internalItem: SeriesMarker<TimePoint>) => {
+			// this doesn't work because of ts-transformer-properties-rename
+			// const { originalTime, ...item } = internalItem;
+
+			// we should use here local variable, otherwise ts-transformer-properties-rename renames the field time to _internal_time
```
[the remark this answers is not in the document]
Comment (kirchet) [reply src/api/series-api.ts]: https://github.com/tradingview/lightweight-charts/pull/934/commits/f66ff411f50ad31585a491930410ca2f63f32436#diff-b5e5ef864e4f34adce2f659e3061587abfa34593384a15e68f48e196bf60e530R152-R156

File: src/model/series-data.ts [hunk 4]
```diff
@@ -16,6 +16,7 @@ export interface SeriesPlotRowTypeAtTypeMap {
 }
 
 export type SeriesPlotRow<T extends SeriesType = SeriesType> = SeriesPlotRowTypeAtTypeMap[T];
+export type SeriesPlotRowWithWhitespaces<T extends SeriesType = SeriesType> = SeriesPlotRowTypeAtTypeMap[T] | WhitespacePlotRow;
```
[the remark this answers is not in the document]
Comment (kirchet) [reply src/model/series-data.ts]: https://github.com/tradingview/lightweight-charts/pull/934/commits/f66ff411f50ad31585a491930410ca2f63f32436#diff-d8fd74e8ac6eb3df9f41ed4c2aca57a3065b4f4beb1eb21961e94c48021b4b0dL19

File: src/model/time-scale.ts [hunk 5]
```diff
@@ -14,13 +14,12 @@ import { defaultTickMarkFormatter } from './default-tick-mark-formatter';
...
 import { areRangesEqual, RangeImpl } from './range-impl';
-import { TickMarks } from './tick-marks';
+import { TickMark, TickMarks } from './tick-marks';
 import {
-	BusinessDay,
 	Logical,
 	LogicalRange,
 	SeriesItemsIndexesRange,
-	TickMarkWeight,
+	TickMarkWeight, Time,
```
[the remark this answers is not in the document]
Comment (kirchet) [reply src/model/time-scale.ts]: https://github.com/tradingview/lightweight-charts/pull/934/commits/f66ff411f50ad31585a491930410ca2f63f32436#diff-fefb9f6a6e16812458e111078af1268f46c22a5bdc162e5e3a1e79562ace6a5fR22-R23

File: src/api/series-api.ts [hunk 6]
```diff
@@ -126,16 +127,39 @@ export class SeriesApi<TSeriesType extends SeriesType> implements ISeriesApi<TSe
...
 			...marker,
+			originalTime: marker.time as unknown as OriginalTime,
 			time: convertTime(marker.time),
 		}));
 		this._series.setMarkers(convertedMarkers);
 	}
 
+	public markers(): readonly SeriesMarker<Time>[] {
+		return this._series.markers().map<SeriesMarker<Time>>((internalItem: SeriesMarker<TimePoint>) => {
+			// this doesn't work because of ts-transformer-properties-rename
```
Comment (timocov) [inline src/api/series-api.ts:153]: it still doesn't work? 🤔

Comment (kirchet) [reply src/api/series-api.ts]: Seems that it's working now
Fixed: https://github.com/tradingview/lightweight-charts/pull/934/commits/f6e24e6637298194f22c63a768232d3243b6aeac

**Comment by timocov** [comment]
Related to https://github.com/tradingview/lightweight-charts/issues/938

**Review by timocov — APPROVED** [review]

**Pull request merged**